Refacto layer; warning big PR :)

peppsac opened the pull request:
This PR is quite big, but it's not intended to be merged as is. I want to iterate on this work but it'll be faster if you can all give early feedback.

Here's the wording from the most important commit:
```
refactor: deep rework of layers and threejs integration
    
    See issues #217, #124, #40 for related discussion.
    
    This massive commit:
      * unifies layers handling (geometry layers are no longer special citizen)
      * split processing in smaller units: these units will be more easy to reuse. For
    example implementing Plane support will be easier because we're now able to share
    the Tile and LayeredMaterial concepts without being forced to use Globe based
    concept (e.g horizontalCulling).
      * limits the duplication of feature between three.js and iTowns: we get the best
    of three.js and we remove incoherencies (e.g our Node were slightly different)
      * breaks the Globe|Quadtree centric vision of iTowns to ease the implementation of
    other types of vizualisaiotn (oriented images, plane mode, pointcloud, etc)
```

The last 7 commits are from WFS part 1 PR #195 and are just included as an example.

peppsac committed:
refactor: deep rework of layers and threejs integration

See issues #217, #124, #40 for related discussion.

This massive commit:
  * unifies layers handling (geometry layers are no longer special citizen)
  * split processing in smaller units: these units will be more easy to reuse. For
example implementing Plane support will be easier because we're now able to share
the Tile and LayeredMaterial concepts without being forced to use Globe based
concept (e.g horizontalCulling).
  * limits the duplication of feature between three.js and iTowns: we get the best
of three.js and we remove incoherencies (e.g our Node were slightly different)
  * breaks the Globe|Quadtree centric vision of iTowns to ease the implementation of
other types of vizualisaiotn (oriented images, plane mode, pointcloud, etc)

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -144,6 +166,7 @@ ApiGlobe.prototype.addImageryLayer = function addImageryLayer(layer) {
 
 ApiGlobe.prototype.addImageryLayerFromJSON = function addImageryLayerFromJSON(url) {
     return Fetcher.json(url).then((result) => {
+        result.update = updateLayeredMaterialNodeImagery;
```
already in `addImageryLayer`

peppsac replied:
Yes, this change was temporary.
The arbitrary split between elevation|imagery|geometry will be removed too.

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -155,25 +178,78 @@ ApiGlobe.prototype.addImageryLayerFromJSON = function addImageryLayerFromJSON(ur
+    for (const url of urls) {
+        proms.push(
+            Fetcher.json(url).then((layer) => {
+                layer.update = updateLayeredMaterialNodeImagery;
```
`layer.update = updateLayeredMaterialNodeImagery;`
already in `addImageryLayer`

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -155,25 +178,78 @@ ApiGlobe.prototype.addImageryLayerFromJSON = function addImageryLayerFromJSON(ur
+        proms.push(
+            Fetcher.json(url).then((layer) => {
+                layer.update = updateLayeredMaterialNodeImagery;
+                layer.graftedOnId = parentId;
```
why `graftedOnId`? 
And not 
```js
layerParent.add(layer)
```

gchoqueux replied:
```js
layer.graftedOnId = parentId;
// It's very similar
layerParent.add(layer);
```

gchoqueux replied:
What do you think ?

peppsac replied:
This code has changed. Now you can do:
```js
let scene = new Scene(...);
scene.addLayer(tile);
scene.addLayer(imagery, tile.id);
```
(see index_plane.html: https://github.com/iTowns/itowns2/pull/239/files#diff-912a8a195bb3f6b6ae7543308f9fe79fR96)

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -360,22 +457,109 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+        nodeType: TileMesh,
+        maxLevel: 16,
+        ellipsoid,
+        initRootNodes: initTiledGeometryLayer,
```
why call `initRootNodes` in `updateTreeLayer`?
It 's possible to call before the first  `updateTreeLayer`?

gchoqueux commented on src/Process/TiledNodeProcessing.js:
```diff
@@ -0,0 +1,117 @@
+            requestNewTile(context.scheduler, layer, layer.schemeTile.getRoot(i), undefined, 0));
+    }
+    Promise.all(_promises).then((roots) => {
+        layer.root = roots;
```
why keep `layer.root`?

peppsac replied:
`root` is a bad name here, maybe I should call this `level1Tiles` - this name is internal to `TiledNodeProcessing` anyway.

gchoqueux replied:
what is the difference between `layer.level1Tiles` and the old `layer.children`

peppsac replied:
Explicitness.

gchoqueux replied:
what is the difference between `layer.root` and the old [`layer.children`](https://github.com/iTowns/itowns2/blob/master/src/Scene/Node.js#L12) and [rootNode](https://github.com/iTowns/itowns2/blob/master/src/Scene/Quadtree.js#L28)

peppsac replied:
`rootNode` served no purpose. It could have been a `THREE.Group` or simply removed, which is was I did.
`layer.children` forced a tree structure for layers. But all layers don't need a tree: actually we have only 1 layer that currently uses a tree structure (tile).
And I renamed `root` because it's a generic name and I believe `level1Tiles` is more clear/explicit.

gchoqueux replied:
if layer have tree structure, The usual name is  `children`

peppsac replied:
When you implement a generic tree structure yes.
But here we're implementing a tree structure holding tiled / subdividing nodes, so I still think that `level1Tiles` is a better name.
Furthermore layers and their content do not have a parent/children relationship anymore. The globe layer simply holds a tree-data structure. So if we were to choose a generic name I'd go with `root`.

Jeremy-Gaillard replied:
I would use `level0Tiles`, I think it is a more common name the root nodes.

I agree with @peppsac on this issue. The layer's content is not the layer's child.

peppsac replied:
Ah indeed, they are level0 tiles, I'll change the name. Thanks

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -360,22 +457,109 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+        cullingTest: globeCulling,
+        preUpdate: (context, layer) => preGlobeUpdate(context, layer),
+        initNewNode,
+        mustSubdivide: globeSubdivisionControl,
```
It's possible to chain processes, next process is called if the previous return true
```js
{
treeProcess: [preUpdate, cullingTest, mustSubdivide],
}
```

peppsac replied:
Because currently `processTiledGeometryNode` is slightly more complicated that simply chaining 3 functions.

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -360,22 +457,109 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+        id: globeLayerId,
+        update: (context, layer, node) => updateTreeLayer(processTiledGeometryNode, context, layer, node),
+        nodeType: TileMesh,
+        maxLevel: 16,
```
`maxLevel` must be linked with imagery and elevation layers
if imagery or elevation are superior

peppsac replied:
Can you explain a bit more?

gchoqueux replied:
Why maxlevel is 16?

peppsac replied:
It's a typo, I'll restore the original value (18)

gchoqueux replied:
globe's maxLevel must be greater than or equal to the maxLevel of children layers

gchoqueux replied:
What do you think ?

peppsac replied:
I think that it's the user responsability to handle this.

gchoqueux replied:
I think by default, It is our responsibility

peppsac replied:
What about implementing this in the globe layer `preUpdate` function?

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -360,22 +457,109 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+        schemeTile: globeSchemeTileWMTS(globeSchemeTile1),
+        cullingTest: globeCulling,
+        preUpdate: (context, layer) => preGlobeUpdate(context, layer),
+        initNewNode,
```
`initNewNode` Is this the right place?

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -360,22 +457,109 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
-    this.scene.scheduler.addProtocolProvider('wmtsc', wmtsProvider);
-    this.scene.scheduler.addProtocolProvider('tile', new TileProvider(ellipsoid));
-    this.scene.scheduler.addProtocolProvider('wms', new WMS_Provider({ support: map.gLDebug }));
+ApiGlobe.prototype.createScene = function createScene(viewerDiv, bbox) {
```
`createScene` and `createSceneGlobe` must be factorize

peppsac replied:
The shared code between `createScene*` methods should probably go in Scene's constructor or a `init` method.
And `createScene` should be removed from ApiGlobe because it's not globe-specific.

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -138,18 +130,54 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+    const config = this.layersConfiguration;
+
+    // parent-free layers (hack for now)
+    const parentFreeLayers = config.getGeometryLayers().filter(l => !l.graftedOnId);
```
[This joins the commentary](https://github.com/iTowns/itowns2/pull/239/files#r97976160)  `layerParent.add(layer)`
```js
let graftedLayers = config.getColorLayers().filter(l => l.graftedOnId === layer.id);
graftedLayers = graftedLayers.concat(config.getElevationLayers().filter(l => l.graftedOnId === layer.id));
graftedLayers = graftedLayers.concat(config.getGeometryLayers().filter(l => l.graftedOnId === layer.id));
```
comes
```js
const cLayers = layer.children;
```

peppsac replied:
I didn't want to use the parent/child semantics to make sure there's no confusion with the scenegraph parent/child relationship.
But I agree that this part must be simplified.

gchoqueux replied:
This is the same mechanism? and The use of children is much simpler

peppsac replied:
You're looking at an old version of the code.
See: https://github.com/iTowns/itowns2/pull/239/files#diff-a6b95dde1d808f9cea55c3fc36fc7f2bR159

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -138,18 +130,54 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+        // layer update (aka: node.layer)?
+
+        // then update for real
+        for (let i = 0; i < elements.length; i++) {
```
Why not update element in `layer.update(context, layer).forEach(e => elements.push(e))`
Ah ok, I understand, you need to push `newElements`.

This seems to me a little complicated and not generic enough

peppsac replied:
Seems generic to me actually: at each step the layer returns the elements he wants to get updated 
It's similar to BrowseTree that kept updating if the process() function returned true but without forcing to use a parent/children structure.

But if you have something simpler I'm all ears :)

gchoqueux replied:
I will look at it

gchoqueux commented on src/Process/TreeLayerProcessing.js:
```diff
@@ -0,0 +1,10 @@
+export default function updateTreeLayer(processNodeFn, context, layer, node) {
+    if (!node) {
```
it's possible to move `initRootNodes` and so remove `updateTreeLayer`

peppsac replied:
With `initRootNodes` and `updateTreeLayer` as they are, itowns provide a framework to implement a tree-based layer: you just need to add these 2 properties and you're done.

If the `initRootNodes` must be done by the user, you transfer him responsabilities: the user must implement the func *and* decide when to call it (before or after scene creation ? before or after other layers creations ? etc etc)

gchoqueux replied:
The framework may have the responsibility of `initRootNodes` when creating the layer

peppsac replied:
If we don't force the timing by controlling when the method is called, I'll leave it completely to the user to make sure that its layer is fully ready.
I'll give it a try.

gchoqueux commented on src/Core/Commander/Providers/TileProvider.js:
```diff
@@ -46,21 +39,28 @@ TileProvider.prototype.executeCommand = function executeCommand(command) {
-    parent.worldToLocal(params.center);
+    if (parent) {
+        parent.worldToLocal(params.center);
+        // TODO: move this elsewhare
```
+1 to move

gchoqueux commented on src/Globe/TileMesh.js:
```diff
@@ -124,20 +100,70 @@ TileMesh.prototype.disposeChildren = function disposeChildren() {
-    for (var material of this.materials) {
-        material.visible = show;
+TileMesh.prototype.childrenLoaded = function childrenLoaded() {
+    // TODO: '4' is specific to Quadtree
```
+1 to move

peppsac replied:
Note that `childrenLoaded` will be removed anyway

gchoqueux commented on src/Scene/Quadtree.js:
```diff
@@ -1,144 +0,0 @@
-/**
```
why remove quadtree?

peppsac replied:
Because it's useless.
What purpose does it serve?

gchoqueux replied:
ok

gchoqueux commented on src/Process/TiledNodeProcessing.js:
```diff
@@ -0,0 +1,117 @@
+        for (const root of roots) {
+            // TODO: support a layer.root attribute, to be able
+            // to add a layer to a three.js node
+            context.scene.gfxEngine.scene3D.add(root);
```
Even logical than the old `Layer.js`?

peppsac replied:
Sort of.
For this layer it looks similar. Other layers (eg wfs) doesn't have any layer.root attribute (nor need them).

gchoqueux reviewed (commented):
Good work!
I make a first review ...
This PR follows the direction of (#217) treeProcess and multiTreeData

peppsac commented:
Made a first update of the PR, fixing some of the comments and a few TODOs - mostly cleanups.

peppsac commented:
I've also moved some code between `ApiGlobe` and `Scene`.
Functions tied to the Globe (e.g everything using the ellipsoid) should now live in `ApiGlobe`.

I believe `Scene` should be exposed to cover users wanting of itowns that don't need a globe.
We could also rename `Scene` to avoid the confusion with three.js `Scene`.

This class holds everything needed to display something using iTowns (layers, camera, controls), so my best proposal is: `Viewer`.
Another possibility is `Map` (similar to Leaflet naming scheme) but iTowns can do way more than 2D maps (pointcloud, panorama, mobilemapping, etc) so...

gchoqueux commented:
> This class holds everything needed to display something using iTowns (layers, camera, controls), so my best proposal is: `Viewer`.

+1 for Viewer

gchoqueux commented on examples/layers/JSONLayers/FL_Buildings.json:
```diff
@@ -0,0 +1,19 @@
+    "title"     : "BDTOPO_BDD_WLD_WGS84G:bati_remarquable,BDTOPO_BDD_WLD_WGS84G:bati_indifferencie,BDTOPO_BDD_WLD_WGS84G:bati_industriel",
+    "styleName" : "",
+    "extractStyle" : true,
+    "level"     : 16,
```
`level` and `bbox` are wfs protocol options?

peppsac replied:
I don't know, but the layer can hold options not needed by the protocol.
In this case, level and bbox are used by `FeatureProcessing`.

gchoqueux replied:
Yes, but it gives the responsibility to the user and this should be the case

peppsac replied:
bbox could be retrieved using  `GetCapabilities`.
But `level` is an (optional) information that the user can give to iTowns - I don't see how iTowns could make that choice on itself (I'm not talking about the specific case of the globe, but iTowns as a framework).
Maybe the API could support attaching to another layer at specific levels:
```js
scene.addLayer(layer, parentLayer, 16)
scene.addLayer(layer, parentLayer, [16, 17, 18])
```
But of course it only makes sense if parentLayer is based on levels...

peppsac commented:
Another update, getting closer to something that can be merged.
I've removed all the WFS stuff, the planar mode for now. They'll get back in with their own PR.

Take a look at `src/Main.js` because I'm exposing more itowns' classes (à la three.js) to allow a user to build a complete itowns-based applications without relying on npm or a loader.

Please voice your concerns on the changes being done here because I'd really like to move this forward and get it merged soon...

(note: the first 7 commits are small fixes that could be merged in their own PR, I'll probably split them in the next iteration)

peppsac commented:
This PR gets smaller and smaller as some commits were extracted to #251 #252 ...
Should be easier than ever to review, so give it a try :)

Jeremy-Gaillard commented on examples/layers/JSONLayers/Region.json:
```diff
@@ -7,7 +7,9 @@
-	"bbox"      : [-61.80983300000002, -21.38963079935857, 55.83665389999998, 51.0890011990783],
+    "unit"      : 1,
+	"bbox"      : [-61.80983300000002, 55.83665389999998, -21.38963079935857, 51.0890011990783],
+    "bbox_url"  : "swne",
```
Mixed space and tabs

Jeremy-Gaillard commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -43,137 +49,141 @@ const defer = function defer() {
- * This function adds an imagery layer to the scene. The layer id must be unique. The protocol rules wich parameters are then needed for the function.
+ * Add an elevation layer to the map. Elevations layers are used to build the terrain.
+ * Only one elevation layer is used, so if multiple layers cover the same area, the one
+ * with best resolution is used (or the first one is resolution are identical).
```
or the first one **if the** resolution**s** are identical

peppsac replied:
These typo are already in master so I won't fix them in this PR (even if it would be a good idea to fix them at some point).

Jeremy-Gaillard commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -43,137 +49,141 @@ const defer = function defer() {
+ * Add an elevation layer to the map using a JSON file.
+ * Elevations layers are used to build the terrain.
+ * Only one elevation layer is used, so if multiple layers cover the same area, the one
+ * with best resolution is used (or the first one is resolution are identical).
```
or the first one **if the** resolution**s** are identical

Jeremy-Gaillard commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -43,137 +49,141 @@ const defer = function defer() {
+ * Add an elevation layer to the map using an array of JSON files.
+ * Elevations layers are used to build the terrain.
+ * Only one elevation layer is used, so if multiple layers cover the same area, the one
+ * with best resolution is used (or the first one is resolution are identical).
```
or the first one **if the** resolution**s** are identical

Jeremy-Gaillard commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -43,137 +49,141 @@ const defer = function defer() {
-    this.scene.getMap().layersConfiguration.moveLayerUp(layerId);
-    this.scene.getMap().updateLayersOrdering();
+    this.scene.layersConfiguration.moveLayerUp(layerId);
+    this.scene.updateLayersOrdering();
```
Scene.updateLayersOrdering() isn't implemented

Jeremy-Gaillard commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -353,56 +303,160 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+        processNode: processTiledGeometryNode,
+        // options for tilegeometry
+        maxLevel: 18,
+        schemeTile: globeSchemeTileWMTS(globeSchemeTile1),
```
Is it a good time to correct weird English?

schemeTile -> tileScheme

peppsac replied:
Not really; this PR is big enough :)

Jeremy-Gaillard commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -434,8 +490,8 @@ ApiGlobe.prototype.orbit = function orbit(value) {
-    this.scene.getMap().setLayerOpacity(id, opacity);
-    this.scene.renderScene3D();
+    this.scene.layersConfiguration.setLayerAttribute(id, 'opacity', opacity);
+    this.scene.notifyChange(0, true);
```
Why is notifyChange called here when renderScene3D is used in previous functions?

peppsac replied:
No real reason. `Scene` should probably expose only 1 method to update its content/rendering.

Jeremy-Gaillard commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -556,7 +612,7 @@ ApiGlobe.prototype.getRange = function getRange() {
     // TODO: error is distance is big with ellipsoid.intersection(ray) because d < 0
     var controlCam = this.scene.currentControls();
-    var ellipsoid = this.scene.getEllipsoid();
+    var ellipsoid = this.ellipsoid;
```
Seems pointless since ellipsoid is used once in the entire function.

peppsac replied:
Indeed, removed.

Jeremy-Gaillard commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -42,37 +40,45 @@ WMS_Provider.prototype = Object.create(Provider.prototype);
-        mE.arrayDegToRad(layer.bbox);
-        layer.bbox = new BoundingBox(layer.bbox[0], layer.bbox[2], layer.bbox[1], layer.bbox[3]);
+    if (!layer.name) {
+        throw new Error('layerName is required.');
```
Inconsistency: layerName -> name

peppsac replied:
Agreed - but this commit is in PR #253 so I'll make the change there.

Jeremy-Gaillard commented on src/Globe/TileMesh.js:
```diff
@@ -213,7 +205,7 @@ TileMesh.prototype.setTexturesLayer = function setTexturesLayer(textures, layerT
         this.material.setTexturesLayer(textures, layerType, layer);
     }
-    this.loadingCheck();
+    // this.loadingCheck();
```
Remove?

peppsac replied:
Done

Jeremy-Gaillard commented on src/Globe/TileMesh.js:
```diff
@@ -183,9 +157,27 @@ TileMesh.prototype.setTextureElevation = function setTextureElevation(elevation)
     this.materials[RendererConstant.ID].uniforms.texturesCount.value = this.materials[RendererConstant.FINAL].loadedTexturesCount[0];
 
-    this.loadingCheck();
+    // this.loadingCheck();
```
Remove?

peppsac replied:
Done

Jeremy-Gaillard commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -20,7 +20,7 @@ emptyTexture.level = EMPTY_TEXTURE_LEVEL;
 var vector = new THREE.Vector3(0.0, 0.0, 0.0);
 var vector4 = new THREE.Vector4(0.0, 0.0, 0.0, 0.0);
-var showDebug = false;
+var showDebug = true;
```
Switch to false before merging

peppsac replied:
Yup

Jeremy-Gaillard commented on src/Scene/Scene.js:
```diff
@@ -7,76 +7,139 @@
     this.needsRedraw = false;
     this.lastRenderTime = 0;
     this.maxFramePerSec = 60;
+    this._extents = new THREE.Box3(new THREE.Vector3(0, 0, 0), new THREE.Vector3(1, 1, 1));
```
This doesn't seem to be used any more.

If you still want to keep it, rename to either _extends ("the scene extends...") or _extent ("the scene's extent). I favours the second option.

peppsac replied:
Removed

Jeremy-Gaillard commented on src/Scene/Scene.js:
```diff
@@ -7,76 +7,139 @@
 Scene.prototype.constructor = Scene;
+
+
+Scene.prototype.initProviders = function initProviders() {
```
Not a fan of this function. I would prefer initialising the providers in the API (or index), so we can easily add new protocols without modifying the core. Also, it adds dependencies to the various provider.

peppsac replied:
You can add providers from `index.html`.
This function only initializes the default providers (and you are free to *not* call it if you don't want them)

Jeremy-Gaillard replied:
Okay, so maybe rename it to initDefaultProviders?

I still think it's weird that this function is in Scene, but I'm not sure there's a better place for it now.

peppsac replied:
Yep I'll do this.
Note that `Scene` is going to have a new name soon; right now `Viewer` is the prefered choice but I'm open for other suggestions (see https://github.com/iTowns/itowns2/pull/239/#issuecomment-276332645)

Jeremy-Gaillard commented on src/Scene/Scene.js:
```diff
@@ -7,76 +7,139 @@
-
-    this.cameras = null;
     this.selectNodes = null;
     this.scheduler = Scheduler(this);
```
This is old code, but the scheduler seems out of place.

I would instantiate it elsewhere and add it to the context.

Jeremy-Gaillard reviewed (dismissed):
I noticed two bugs related to picking:
- You can't pick tiles that are not directly in front of the camera
- After picking (or tying to pick) an object, the first mouse drag does not work

peppsac commented:
@Jeremy-Gaillard regarding picking: the issue here is that atmosphere mesh can be selected instead of the underlying tiles.
The Picking API basically needs to allow the user to perform picking only on some layers, because sometimes you want to select only the globe (e.g for camera controls), sometimes you want to pick only building, etc etc
(you can see in `c3DEngine.prototype.renderTobuffer` that I already implemented something similar, but in a hardcoded way)

gchoqueux commented on examples/GUI/GuiTools.js:
```diff
@@ -35,14 +35,14 @@ GuiTools.prototype.addImageryLayerGUI = function addImageryLayerGUI(layer) {
     });
     folder.add({ frozen: false }, 'frozen').onChange((value) => {
-        this.api.scene.getMap().layersConfiguration.setLayerFreeze(layer.id, value);
+        this.api.scene.layersConfiguration.setLayerFreeze(layer.id, value);
```
How do you manage multiple map in a page?
One map by scene?

The layer are attached to tiles layers?
```js 
const imageryPromise = itowns.viewer.addImageryLayersFromJSONArray(urlImageryLayers, globeLayer.id);
```

gchoqueux commented on index.html:
```diff
@@ -60,15 +56,12 @@
-            itowns.viewer.addGeometryLayer(wgs84TileLayer);
-            const imageryPromise = itowns.viewer.addImageryLayersFromJSONArray(urlImageryLayers);
-            const elevationPromise = itowns.viewer.addElevationLayersFromJSONArray(urlElevationLayers);
+            const imageryPromise = itowns.viewer.addImageryLayersFromJSONArray(urlImageryLayers, globeLayer.id);
```
Add default `parentLayer` if only one map

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -43,137 +49,156 @@ const defer = function defer() {
-ApiGlobe.prototype.add = function add(/* Command*/) {
-    // TODO: Implement Me
-
+ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer, parentLayerId) {
```
no default `parentLayerId`

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -43,137 +49,156 @@ const defer = function defer() {
+    layer.update = updateLayeredMaterialNodeElevation;
+
+    this.scene.addLayer(layer, parentLayerId);
+    this.scene.layersConfiguration.setLayerAttribute(layer.id, 'type', 'elevation');
```
why you typed layer?

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -353,56 +318,160 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
-    this.scene.scheduler.addProtocolProvider('wmtsc', wmtsProvider);
-    this.scene.scheduler.addProtocolProvider('tile', new TileProvider(ellipsoid));
-    this.scene.scheduler.addProtocolProvider('wms', new WMS_Provider({ support: map.gLDebug }));
+    this.scene.initDefaultProviders(this.scene);
```
strange :  `this.scene.initDefaultProviders` take `this.scene` in parameter

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -353,56 +318,160 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+        // layer base options
+        protocol: 'tile',
+        id: globeLayerId,
+        preUpdate: (context, layer) => preGlobeUpdate(context, layer),
```
`preGlobeUpdate` Can be named `preUpdateTreeLayer`
or `preGlobeUpdate` is specific toh the Globe?

peppsac replied:
Yes it's globe-specific because it prepares horizon culling.

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -137,19 +191,60 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+    const context = {
+        camera: this.camera,
+        scheduler: this.scheduler,
+        scene: this,
```
context have `scene` to notify change?

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -353,56 +318,160 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+
+    this.addGeometryLayer(wgs84TileLayer);
+
+    const atmosphere = new Atmosphere(this.ellipsoid);
```
instance atmosphere here, is to get out all the specificities of the globe, isn't?

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -353,56 +318,160 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+    this.layers[0].node.updateLightingPos(lightingPos);
+};
+
+ApiGlobe.prototype.animateTime = function animateTime(value) {
```
`animateTime` I'm not sure it's still working

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -353,56 +318,160 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+    }
+};
+
+ApiGlobe.prototype.orbit = function orbit(value) {
```
orbit : I'm not sure that it still serves

peppsac replied:
If it's useless I can remove it too

gchoqueux commented on src/Globe/TileMesh.js:
```diff
@@ -66,77 +67,50 @@ function TileMesh(params, builder, geometryCache) {
 };
 
-TileMesh.prototype.setUuid = function setUuid(uuid) {
-    this.id = uuid;
```
select tile is broken

peppsac replied:
Should be fixed, can you confirm?

gchoqueux replied:
yes, I will prepare PR from this [commit](https://github.com/iTowns/itowns2/pull/214/commits/86d76ec5ca017b6ce3935b67c23363c29aa2eb71) where I could fix select tile

gchoqueux replied:
rework

gchoqueux commented on src/MobileMapping/MobileMappingLayer.js:
```diff
@@ -1,148 +0,0 @@
-/**
```
You have totally removed MobileMappingLayer.js ?
If this is the case it is very good +1
It will come back at the right time

peppsac replied:
I removed all Layer(s) instances.
And as MobileMapping feature was broken I didn't port it to the new arch.

gchoqueux commented on src/Renderer/PanoramicMesh.js:
```diff
@@ -6,11 +6,11 @@
  */
```
remove `PanoramicMesh.js` ?

gchoqueux commented:
* No blue sky
 * Realistic lighting is broken : you wait the PR lighting?

peppsac commented:
Fixed a few things and rebased the PR on master.
It was a bit painful to say the least...

Now that PR #248 is merged it's a good time to think about the content of `Main.js` (ie: what do we want to expose directly to html).
I don't like the current version of `Main.js` in my PR, and I think we should move toward exposing directly useful classes/methods (like three.js does).

Let me know what you think.

Jeremy-Gaillard commented:
A few bugs:
- Realistic lightning is still broken
- Planar mode camera acts weirdly (its behaviour  doesn't seem consistent)
- Planar mode tiles tend to disappear/reload when rotating the camera
- examples/index.html and examples/helloWorld.html do not work

peppsac commented:
* Fixed realistic lighting
* I'd need more details (though we shouldn't focus much on Planar mode. I'll probably move this code to another PR)
* Yeah, probably an issue with culling. But see comment above.
* Yes, but it's not my fault :)

peppsac commented:
Note : I've reworked `Main.js`, and it's now looking good imho.

```js
export { ApiGlobe } from './Core/Commander/Interfaces/ApiInterface/ApiGlobe';
export { addOBBLayer } from './Renderer/ThreeExtended/OBBHelper';
export { Scene } from './Scene/Scene';
```
I've added a small hack to allow access to three.js and proj4js directly from html (will remove this as soon as we're capable of building a `vendor.js` containing these dependencies).
In the meantime it allows to do this : 
```js
itowns.proj4.defs('EPSG:3946', '....');
```
directly in html, and this EPSG will be available to itowns Coordinates module.

peppsac commented:
Sooooo.... I've split again this PR :)
The 2 commits left are basically covering only 1 topic: layer refactoring.

As soon as we agree to merge this one, I'll submit other PRs:
* one shuffling code between `Scene` and `ApiGlobe` 
* one refactoring itowns external API (`Main.js`)
* and probably other with cleanups or small features

peppsac commented:
Pushed a rename commit.
I'll rename `Scene` -> `Viewer` when we're ready to merge, because it'll make the PR very hard to review...

peppsac commented:
Known issue: camera movement is broken when realisitic lighing is enabled (thanks @Jeremy-Gaillard for noticing).
I'll push a commit fixing this soon, using the new `threejsLayer` attribute to filter out interfering objects, such as atmosphere.

autra commented on src/Scene/SceneConfiguration.js:
```diff
@@ -0,0 +1,151 @@
+    this.layersState = {};
+}
+
+SceneConfiguration.prototype.constructor = SceneConfiguration;
```
Nits: Not needed, since you don't reassign prototype.

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -352,49 +398,91 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
+    var positionCamera = new C.EPSG_4326(coordCarto.longitude, coordCarto.latitude, coordCarto.altitude);
 
-    var map = new Globe(gLDebug);
+    this.scene = Scene(positionCamera, ellipsoidSizes(), viewerDiv, debugMode, gLDebug);
```
At one point, we should do `new Scene`. We need one instance of scene per itowns instance (potentially several per page).

autra commented on src/Globe/TileMesh.js:
```diff
@@ -143,10 +133,26 @@ TileMesh.prototype.setTextureElevation = function setTextureElevation(elevation)
+        return undefined;
+    }
+
+    return (this.parent.level === level) ? this.parent : this.parent.getNodeAtLevel(level);
```
Cosmetic: I would replace from line 146:
```js
if (this.parent.level === level) {
    return this.parent;
} 
if (typeOf(this.parent.getNodeAtLevel) !== 'function') {
    return undefined;
}
return this.parent.getNodeAtLevel(level);

```
I find it more readable.

It works, because typeof(undefined) is 'undefined', typeof(null) is 'object' (yes, even though null does not have \_\_proto__ set to Object... Yeah, javascript...). When used for `Function`, typeof is actually a valid null/undefined test.

autra commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -0,0 +1,133 @@
+    var points = node.OBB().pointsWorld;
+    var isVisible = false;
+
+    for (var i = 0, max = points.length; i < max; i++) {
```
nits: `for (const point of points)`? but maybe you copy-pasted this from elsewhere...

autra commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,309 @@
+    if (parent && parent.material && parent.material.getLayerLevel) {
+        var level = parent.material.getLayerLevel(layerType, layer ? layer.id : undefined);
+        if (level >= 0) {
+            return parent.getNodeAtLevel(level);
```
Are we sure we always have a `getNodeAtLevel` function? Some code above let me think we don't.

autra commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,309 @@
+            const level = ancestor ? ancestor.level : node.level;
+            // Assign .level to texture
+            if (Array.isArray(result)) {
+                for (let j = 0; j < result.length; j++) {
```
Nits: `for..of`?

autra commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,309 @@
+        });
+}
+
+export function updateLayeredMaterialNodeElevation(context, layer, node) {
```
Nits: Let's get rid of the private function if you're just proxying.

peppsac replied:
Well.
The private function returns a promise (which is used in `initNewNode`) but the regular update function should return nothing or an array of elements to update...
I agree this is rather ugly, but it gets the job done for now. Maybe a good candidate for refactoring later though.

autra replied:
Oh ok, all right, that's perfectly acceptable.

autra commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,309 @@
+    const currentElevation = node.materials[RendererConstant.FINAL].getElevationLayerLevel();
+
+    // Step 0: currentElevevation is -1 BUT material.loadedTexturesCount[l_ELEVATION] is > 0
+    // means that we already tried and failed to download an elevation texture
```
^_^ this is fishy ;-) I have nothing better to propose right now though.

autra replied:
I mean, we do have a layerUpdateState right? Please ignore if it comes from a previous version though.

peppsac replied:
Agreed but this is a copy-paste from `NodeProcess.js`.
And it will be improved when PR #244 gets merged.

autra commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,309 @@
+    // We don't care about layer status (isLayerFrozen) or limits (tileInsideLimit) because
+    // we simply want to use ancestor's texture with a different pitch
+    if (currentElevation == -1) {
+        ancestor = findAncestorWithValidTextureForLayer(node, parent || node.parent, l_ELEVATION);
```
There are 3 early return before we really use ancestor. Maybe we can move that down? I don't expect `findAncestorWithValidTextureForLayer` to be very costly, but if I'm not mistaken, we are executing that in a `requestAnimationFrame` so...

peppsac replied:
As said above, I'd prefer not to modify to much the code, because I'll have to rebase over PR #244

autra commented on src/Process/TiledNodeProcessing.js:
```diff
@@ -0,0 +1,160 @@
+}
+
+function subdivideNode(context, layer, node) {
+    if (!node.pendingSubdivision && node.children.filter(n => n.layer == layer.id).length == 0) {
```
Nits: is `node.children` a regular array? If so, let me introduce [Array.prototype.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

autra commented on src/Process/TiledNodeProcessing.js:
```diff
@@ -0,0 +1,160 @@
+
+        const promises = [];
+        const children = [];
+        for (let i = 0; i < bboxes.length; i++) {
```
nits: for..of?

autra commented on src/Process/TiledNodeProcessing.js:
```diff
@@ -0,0 +1,160 @@
+    for (let i = 0; i < node.children.length;) {
+        if (node.children[i].layer === layer.id) {
+            node.children[i].dispose();
+            node.children.splice(i, 1);
```
It's usually faster to allocate a new array for these kind of operations.

peppsac replied:
Probably; though this is a rather unusual operation so speed doesn't matter.

autra commented on src/Process/TiledNodeProcessing.js:
```diff
@@ -0,0 +1,160 @@
+    // remove children
+    for (let i = 0; i < node.children.length;) {
+        if (node.children[i].layer === layer.id) {
+            node.children[i].dispose();
```
Snippets of code I've seen on the internet also call `dispose` on the geometry, the material and/or the texture btw. Is it needed here?

peppsac replied:
This is done in `TileMesh.prototype.dispose`

autra commented on src/Renderer/PanoramicMesh.js:
```diff
@@ -6,11 +6,10 @@
 
-import NodeMesh from './NodeMesh';
-
+import * as THREE from 'three';
```
`import { Mesh } from 'three';` ?

peppsac replied:
Yeah, but the general style in itowns is to `import *` so let's keep it coherent (ie: we can do this but in a dedicated PR :) )

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -481,9 +483,15 @@ function GlobeControls(camera, domElement, engine) {
+
+        return function getPickingPosition(controlsActiveLayers, coords)
         {
+            const prev = engineGfx.camera.camera3D.layers.mask;
```
nits: `gfxEngine`?

autra commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,309 @@
+    node.layerUpdateState[imageryLayer.id].texturesCount = texturesCount;
+}
+
+export function initNewNode(context, layer, parent, node) {
```
I'm surprised this function knows about elevation and color layers. Shouldn't all the process of passing elements to children layer update function allow us to avoid that?

autra commented on src/Scene/Scene.js:
```diff
@@ -136,19 +118,60 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+    for (const tree of config.layerTrees) {
+        const elements = [];
+        const layer = tree.layer;
+        // level 0 layers get a first element-less call
```
I'm sure that with some effort, we can get rid of this loop and directly use updateElements.

For a start we can:
- move the layer.initLevel0Nodes call in TreeLayerProcessing.js into the preUpdate fn of the globe layer (and maybe get rid of TreeLayerProcessing.js by getting level0Nodes directly here).
- Then having a real recursion from the start.

The thing is... this smells funny. Alongside updateElement above (which looks like it wants to be the recursive version of the code below) and the weird initialization code in TreeLayerProcessing.js (inside the if (layer.level0Nodes === undefined)), it looks like either things are not initialized where they should (and we struggle here for this reason) or I didn't understand something (in which case consider adding more comments?).

Again, maybe I misunderstood something, but it smells funny to get a first version of elements array, then having to add stuff in it while iterating it, but only doing so for the top layer of the config tree... 

Maybe nomenclature can be improved? Are elements always 3D objects? in that case, let's call them sceneObjects or something more specific. Maybe layer can be called `source` or `dataSource` (I don't really like `process`)? And what is `grafted`?

On overall, I'm pretty sure this method needs more comments. After all, this is not an "easy-to-delete" piece of code anyway, it could be heavily commented without fear.

To summarize, this code is now the core of all itowns. It should be at the same time generic, but also present clear "bounds". The concepts behind it must be clear to everyone because understanding it will be crucial to write one's own layer code. That's why I think it needs special attention.

autra replied:
Here is an alternative proposal:

- add `layer.initLevel0Nodes(context, layer)` at the end of `preGlobeUpdate` in `GlobeTileProcessing.js`.

- the update fn of the geometry layer could be change this way:
```diff
-        update: (context, layer, node) => updateTreeLayer(context, layer, node),
+        update: (context, layer, node) => {
+            if (!node) {
+                return layer.level0Nodes;
+            }
+            return layer.processNode(context, layer, node);
+        },
```

- we get rid of `TreeLayerProcessing.js`. Honestly good riddance, who needs a process to traverse a tree when such a traversal is only a recursion?

- In `Scene.prototype.update`, the whole process becomes:
```js
function updateElements(context, source, elements) {
    for (const element of elements) {
        let elems = source.layer.update(context, source.layer, element)
        if (elems) {
            updateElements(context, source, elems);
        }
        if (source.children) {
            for (const s of source.children) {
                const newElements = s.layer.update(context, s.layer, element);
                if (newElements) {
                    updateElements(context, s, newElements);
                }
            }
        }
    }
}

Scene.prototype.update = function update() {
    this.gfxEngine.camera.update();

    // Browse Layer tree
    const config = this.configuration;

    // TODO?
    const context = {
        camera: this.gfxEngine.camera,
        scheduler: this.scheduler,
        scene: this,
    };

    // call pre-update on all layers
    config.traverseLayers((layer) => {
        if (layer.preUpdate) {
            layer.preUpdate(context, layer);
        }
    });

    // update layers
    for (const tree of config.layerTrees) {
        const layer = tree.layer;
        // Initial call to get elements to update.
        const elements = layer.update(context, layer);

        if (elements) {
            updateElements(context, tree, elements);
        }
    }
};
```
processElements becomes slightly more complicated,  but illustrate better what happens imho. The whole thing is easier to read because there isn't any special initialisation case any more.

I think this could even be improved if the initialisation of the first elements are not done by `update`, but either by a `getInitialElements` (if we want to give client code flexibility), or simply with `layer.level0Nodes` (if we want simplicity). That way we won't even need the test in `update` I guess.

autra reviewed (commented):
Apparently this PR is the future, so I figured I'd have a look ;-)

On overall, I really like what you've done here (I refrained from adding some `<3` everywhere though). We gain in both simplicity and power :-) I believe this would make a very strong and powerful basis to iterate from. Thanks for all the effort!

I've left some comments. Most of them are either nits or probably about stuff that were there before. Address them or not as you see fit (and regarding the energy you still have for this PR ;-) ), it's not very important. Some others are concerns or questions, but the most important ones are  2 comments on Scene.js about the update process, with an alternate proposal. Please tell me what you think.

As a side note, mouse navigation is currently broken for me (both on firefox and on chrome). Are you aware of this issue?

gchoqueux commented:
I wait for the feedback following the comments of @autra

peppsac commented:
Thanks for the review @autra 
You're right, most of the nits you observed are "legacy" code I merely moved to other files. I'll fix the easy ones, and leave the more complex out for this PR.

Regarding your last comment on `Scene.update` function: 
- yes it needs more comments, will add some
- I don't like the naming of most of the code here. Your comment reminds me of this unfortunate situation so I'll give this another try.
- thanks for the alternative update function proposal, I'll use it as inspiration :)

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -0,0 +1,132 @@
+    cV.copy(context.camera.position()).divide(radius);
+    vhMagnitudeSquared = cV.lengthSq() - 1.0;
+
+    // pre-sse
```
pre-SSE and SSE isn't specific to the globe

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -0,0 +1,132 @@
+const frustum = new THREE.Frustum();
+const obbViewMatrix = new THREE.Matrix4();
+
+function frustumCullingOBB(node, camera) {
```
frustumCullingOBB isn't specific to the globe

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -0,0 +1,132 @@
+    return !(frustumCullingOBB(node, camera) && horizonCulling(node));
+}
+
+function computeNodeSSE(camera, node) {
```
computeNodeSSE isn't specific to the globe

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -0,0 +1,132 @@
+    return preSSE * (node.geometricError / distance);
+}
+
+export function globeSubdivisionControl(context, layer, node) {
```
globeSubdivisionControl isn't specific to the globe, 
The name isn't correct, more generic is `nodeSubdivisionControl`

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -0,0 +1,132 @@
+        bbox.maxCoordinate._internalStorageUnit = UNIT.RADIAN;
+    }
+    return schemeT;
+}
```
Final comment for `GlobeTileProcessing.js` 
Several process aren't specific to globe : 

- `pre-SSE`
- `frustumCullingOBB`
- `computeNodeSSE`
- `globeSubdivisionControl`

In the concern of genericity and to avoid the specificities and these processes need to move in a generic file, some suggestions of names: 
genericProcessing.js
nodeProcessing.js
object3DProcessing.js

gchoqueux commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,309 @@
+    }
+}
+
+function nodeCommandQueuePriorityFunction(node) {
```
`nodeCommandQueuePriorityFunction` is not specific to the `LayeredMaterialNode`?
I'm not sure that this function is in the right place.

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -136,19 +116,60 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+    const elements = layer.update(context, layer, element);
+
+    if (elements) {
+        for (const element of elements) {
```
For the moment this part isn't used
Can you add somes comments to explain this `function updateElement`, `element` and `elements`
and add a use case for `elements`

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -136,19 +116,60 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+            }
+
+            // append elements to update queue
+            if (newElements) {
```
`if (newElements)` if always true?

Add use case to have element different from TileMesh

It would be necessary to unify the terms `element` and `node`?

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -104,34 +112,45 @@ function preprocessLayer(layer, provider) {
  * Add the geometry layer to the scene.
  */
-ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
+ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer, parentLayerId) {
```
It will be clearer to do 

`layerParent.addGeometryLayer(layer)` than to `api.addGeometryLayer(layer, parentLayerId)`

To deepen the issue
`layerTrees` or `sceneConfiguration`  looks like a `Map` object. 
We could imagine : 
`Map().layerParent.addGeometryLayer(layer)` or
`Map().addGeometryLayer(layer, parentLayerId)`

peppsac commented:
Updated.

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -136,22 +116,62 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+    // call pre-update on all layers
+    config.traverseLayers((layer) => {
+        if (layer.preUpdate) {
+            layer.preUpdate(context, layer);
```
it isn't necessary to pass layer in preUpdate, preGlobeUpdate don't have param layer ? 
There are cases of use where it will be used?
preUpdate use to calculate contextual values, we could perhaps add them to the context?

peppsac replied:
The `layer` param doesn't hurt so we can probably figure out later if it's useful or not.

I'm not against storing things in context, but as most of this PR content I'd rather wait to see how it's really used before doing more modifications.

gchoqueux reviewed (commented):
I'm a little confused with all the terminologies
* Layer
* Node
* Link
* current in traverse link
* LayerHierarchies
* parentLayer
* root
* nodeLevel0
* source
* element

peppsac commented:
Well yes, we have a lot of similar but not quite the same terms (but it's better than before when everything was a layer :-))
Joking aside (and as said yesterday in IRL meeting): I'd prefer to move forward and merge an imperfect PR rather then waiting to find the perfect wording/code splitting/...

gchoqueux commented:
I agree that it takes a PR for cleaning/renaming

> but it's better than before when everything was a layer

What makes reading confusing is that the use of terms is not homogeneous in the code and the comments. There are many similar terms that do not allow the concepts to be clearly seen.

You still have work on this PR?

gchoqueux commented:
Yo can add some comments to all terminologies?

peppsac commented:
I think I'm done. The naming is not stable yet, so I'll add some documentation and comments later

gchoqueux commented:
So, add some comments to concept attached to theses names : 

* Layer
* Link
* LayerHierarchies
* source
* element

It will help everyone and will allow to go faster

gchoqueux commented:
Why separate `layer` and `link`?
what is the point?

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -136,22 +116,62 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+}
+
+Scene.prototype.update = function update() {
+    this.gfxEngine.camera.update();
```
`this.gfxEngine.camera.update();`
it's needed?

gchoqueux commented:
I notice that there are difference in `update loop` practice.

The update of the scene can be twice as long between 2 positions (Especially the one involving the loading of the pyramid of the quadtree). 

Logically, the number of `Scene.step()` calls is also doubled.

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -162,19 +160,24 @@ Scene.prototype.update = function update() {
 };
 
 Scene.prototype.step = function step() {
+    this.currentControls().updateCameraTransformation();
```
it isn't the place, updateControl had to be done when datas are loaded and all transformation is completed.

peppsac replied:
What does `updateCameraTransformation` actually do?

gchoqueux replied:
`updateCameraTransformation` is not a good naming.
it updates camera's target  and camera according to the globe

peppsac replied:
So it's globeControls specific. I'll look into this.

peppsac commented:
> I notice that there are difference in update loop practice.
> The update of the scene can be twice as long between 2 positions (Especially the one involving the loading of the pyramid of the quadtree).
> Logically, the number of Scene.step() calls is also doubled.

I'm afraid I didn't understand any of this :-/
Are you talking about a performance regression?

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -162,19 +160,24 @@ Scene.prototype.update = function update() {
 
 Scene.prototype.step = function step() {
+    this.currentControls().updateCameraTransformation();
+    this.gfxEngine.camera.update();
```
`this.gfxEngine.camera.update()` is not needed. 
camera.update() is already done in gfxEngine

gchoqueux commented:
> performance regression?

Yes, because under the same conditions, the number of requests is more important.

In first time, I thought that it came from dissimilar subdivision conditions, but this is apparently not the cause of the problem

peppsac commented:
Need code pushed. Major changes:
* simplification. The API for layers became:
```js
const geom = new GeometryLayer();
scene.addGeometryLayer(geom);
geom.attach(imagery1);
geom.attach(elevation);
```
* the update loop is simpler too. It's now 2-stages only:
    - first stage = geometry layers (globe, point cloud, etc)
    - second stage = layers that expect to be called when some geometry is visible. Their update() function receive the geometry elements (eg: a tile)
*  the SceneConfiguration.js is gone. Layer properties are all stored in directly in Layers so:
```js
layer.visible = true
layer.opacity = 0.5;
```

peppsac commented:
> Yes, because under the same conditions, the number of requests is more important.

Please be more specific. Network request ? subdivision request ? requestAnimationFrame ?
Also "more important" ? x1.01 ? x1.1 ? x2 ? x10 ? x1000 ?
Last but not least: provide precise reproduction test. Maybe the issue is obvious and I'll find by my own, maybe it's not and we'll both lose time writing comments like this one.

gchoqueux commented:
* Network request are 2x more important
* To reproduce simply and precisely , use friendly function `setRange(50000)` after globe loaded

peppsac commented:
I get the opposite result :)
`master`: 
- 500 network requests until `Globe loaded`
- 700 network requests until `setRange(50000)` is done

`refacto_layer`:
- 100 net requests until `Globe loaded`
- 550 net request until `setRange(50000)` is done

gchoqueux commented:
do test with `node.level =< 2` in   `globeSubdivisionControl` to have the master's condition

peppsac commented:
Pushed a few more commits.
I believe I've fixed all the comments, let me know if I missed something.

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -233,11 +190,74 @@ ApiGlobe.prototype.addElevationLayersFromJSON = function addElevationLayersFromJ
+    return Promise.all(proms);
+};
+
+function updateLayersOrdering(globeLayer, imageryLayers) {
```
`updateLayersOrdering` is not specific to `globeLayer `, but more to `geometryLayer`

peppsac replied:
Currently this methods belongs to `ApiGlobe` so it expects a globeLayer.
I'll change the name anyway.

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -233,11 +190,74 @@ ApiGlobe.prototype.addElevationLayersFromJSON = function addElevationLayersFromJ
+ * @param      {number}  newIndex   The new index
+ */
+ApiGlobe.prototype.moveLayerToIndex = function moveLayerToIndex(layerId, newIndex) {
+    const imageryLayers = this.scene.getAttachedLayers(l => l.type === 'color');
```
there's `getLayerById`, It doesn't work anymore

peppsac replied:
Yep.
I'll push a commit fixing `getLayerById`, `getLayers` and `getLayersAttribution`

gchoqueux commented on src/Scene/Scene.js:
```diff
@@ -131,27 +111,65 @@ Scene.prototype.scheduleUpdate = function scheduleUpdate(forceRedraw) {
+    return result;
+};
+
+function updateElement(context, geometryLayer, element) {
```
```js
function updateElements(context, geometryLayer, elements) {
    if (!elements) {
        return;
    }
    for (const element of elements) {
        // update attached layers
        for (const attachedLayer of geometryLayer._attachedLayers) {
            attachedLayer.update(context, attachedLayer, element);
        }
        // update element
        const newElementsToUpdate = geometryLayer.update(context, geometryLayer, element);
        updateElements(context, geometryLayer, newElementsToUpdate);
    }
}

Scene.prototype.update = function update() {
    const context = {
        camera: this.gfxEngine.camera,
        scheduler: this.scheduler,
        scene: this,
    };

    for (const geometryLayer of this._geometryLayers) {
        const elementsToUpdate = geometryLayer.preUpdate(context, geometryLayer);
        updateElements(context, geometryLayer, elementsToUpdate);
    }
};
```
otherwise, `updateElements` is recursive function, elements are then necessarily hierarchical.
Therefore it would be clear that those are children.
So replace `Elements` by ` Children` or `ChildElement`?

peppsac replied:
"elements are then necessarily hierarchical" => no.
You can use a recursive function on a flat structure: array, linked-list, etc.
I'll keep the structure-agnosotic name.

gchoqueux replied:
Ok

Otherwise, what about the small rewrite `updateElements` and `update`

peppsac replied:
I applied your change, except that I kept the `// update element` before the `// update attached layers`

gchoqueux commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -373,10 +465,19 @@ ApiGlobe.prototype.setRealisticLightingOn = function setRealisticLightingOn(valu
-    this.scene.getMap().setLayerVisibility(id, visible);
-    this.update();
-    eventLayerChangedVisible.layerId = id;
+ApiGlobe.prototype.setLayerVisibility = function setLayerVisibility(layer, visible) {
```
Doc problem :  `@param {id} string.`

peppsac replied:
We agreed that: 
- this whole file is going away (see https://github.com/iTowns/itowns2/issues/272#issuecomment-288035531)
- itowns documentation needs a complete overhaul

So I don't think it's worth nit picking on doc for this particular PR.

gchoqueux replied:
In this PR the documentation has been deleted or is no longer correct.

Even if apiGlobe is deleted and functions are moved, the documentation will always be associated with these functions and must therefore be up to date. 
The doc has been maintained to date and should be maintained for the benefit of all.

gchoqueux commented:
* `removeImageryLayer` doesn't remove
* `moveLayerUp('Ortho') `doesn't up with 2 layers (Ortho and ScanEX)
* `moveLayerDown('ScanEX')` doesn't up with 2 layers (Ortho and ScanEX) after change position camera
* `moveLayerToIndex('Ortho',1)` doesn't work with 2 layers (Ortho and ScanEX)
* Error 404 loading MNT in `Refacto Layer` and not in Master (example zone Annecy)
* Some tiles haven't the good Zoom MNT While their children are correct
* examples/index.html not updated

peppsac commented:
Pushed a commit fixed the first 4 issues (removeImageryLayer, moveLayerXXXX).

Regarding MNT:
* "Error 404 loading MNT":  it's an old bug, but it's invisible on master. We don't test for the actual WMTS boundaries in `WMTS_Provider.prototype.tileInsideLimit`. Therefore it's expected that the elevation layer tries to fetch a MNT texture, even if it's out of bound:
```
http://wxs.ign.fr/.../wmts?LAYER=ELEVATION.ELEVATIONGRIDCOVERAGE&[...]&TILEMATRIX=11&TILEROW=4017&TILECOL=16946
http://wxs.ign.fr/[...]/wmts?LAYER=ELEVATION.ELEVATIONGRIDCOVERAGE.HIGHRES&[...]&TILEMATRIX=14&TILEROW=4017&TILECOL=16946
``` 
So the first request fails with a 404 error and the second succeeds.

* "Some tiles haven't the good Zoom": I suspect it's the same issue. `master` has a complex code that tries to deal with the limitation of having only 1 elevation texture at a time. `refacto_layer` don't. I'm not 100% of this reasonning, but as I can't repro the issue I can only guess.

peppsac commented:
WMTS bounds testing must be done btw, but not in this PR.

gchoqueux commented:
same comments to :
* moveLayerDown('ScanEX') doesn't up with 2 layers (Ortho and ScanEX) after change position camera
* moveLayerToIndex('Ortho',1) doesn't work with 2 layers (Ortho and ScanEX) after change position camera

* There is still a small problem with elevation
  * `master`
![master](https://cloud.githubusercontent.com/assets/11291849/24258026/d21a7644-0fec-11e7-9346-f72fe71901b4.png)
  * `Refacto layer`
![rl](https://cloud.githubusercontent.com/assets/11291849/24258033/d72169cc-0fec-11e7-94d2-e120c7f43625.png)

gchoqueux commented:
Bug with wireframe after move camera

peppsac commented:
Elevation problem should be fixed.

The 3 other problems won't get fixed until I get precise reproduction steps: 
* what did you **do** 
* what **happened**
* what result did you **expect**

Without this I'm playing a guessing game.
And yes it's more work for you but as a reviewer it's part of your duty.

peppsac commented:
> The 3 other problems won't get fixed

Meh, actually I fixed them.

gchoqueux commented:
Protocol Test to show 2 bugs in WMTS loading.

**Prepare** 
* only 'Ortho' for imagery and 2 DTMs
* enable show outline;
* for master, change this.maxLevel to 18 in Quadtree.js

**Test : in console execute theses commands after globe loaded** 

1. `itowns.viewer.setCenter({longitude:6.5,latitude:46})` : **ok**
2. `itowns.viewer.setRange(500000)` : **Problem in current PR**
      **bad strategy loading : WMTS Ortho color passes by precise level, then degraded, then more precise**
3. `itowns.viewer.setRange(5000)` : **ok**
4. `itowns.viewer.setTilt(66)` : **ok**
5. `itowns.viewer.setRange(500)` : **Problem in current PR**
     **bad DTMs** (see final result)

**Final result** 

  - `Master`
![master](https://cloud.githubusercontent.com/assets/11291849/24301244/64bbd188-10af-11e7-8243-4b1108218ccc.png)
  - `Current PR`
![bad mnt](https://cloud.githubusercontent.com/assets/11291849/24301253/6f45ab60-10af-11e7-8d1a-09d902a54dce.png)

peppsac commented:
Thanks for the repro steps.
After experimenting for quite a while I believe that all these bugs existed in master, but were hidden.

Right now the code of LayeredMaterialNodeProcessing is mostly a copy-paste from master's code but behave differently because the architecture changes that modifies a lot of stuff but mostly :
* timings
* layers dependencies

I have evidence that timing issue cause some blinking imagery layers bugs (my plan is to get them fixed by making the ancestor fetch synchronous instead of asynchronous) and the layer dependencies causes the elevation issue.

But it shouldn't block this PR IMHO. This PR is about moving code in new files and architectural changes.
The fact that old bugs are now visible is irrelevant.

So I'd like to merge this PR, and after that we'll be able to land the PRs depending on it *and* PRs fixing these bugs.

gchoqueux commented:
* The test protocol reveals, in the first place, display bugs
Invisible display bugs, do they exist?

> By making the ancestor synchronous

I have to check but Remove Ancestor PR, but done the job. I propose to merge this before PR, it will help a lot to this PR, which was originally planned.

* These bugs must to be fixed so I do not see the interest of fixing them later. Moreover, they cause regression.

peppsac commented:
Ok, makes sense too.

I'll get back to this PR as soon as possible, probably mid-next week.

vpicavet commented:
Hi,

I understand that this PR let us see bugs that were hidden before.

But I really would like to see this merged as soon as possible : 
- a lot of other work depend on it, it is currently a big blocker for others
- it does not introduce new bug
- we will have little time to fix the existing bugs in the next days

We need to move forward faster, and iterate more. I therefore propose : 
- merge this PR ASAP
- create specific issues for the problems that showed up with this PR
- address these issues in followup PRs

Separation of concerns and PR atomicity is important. Can we unblock the situation ?

gchoqueux commented:
> I understand that this PR let us see bugs that were hidden before

At first glance, [with the protocol of tests,](https://github.com/iTowns/itowns2/pull/239#issuecomment-289057033) it's difficult to affirm that this PR does not introduce a bug :

  - `Master`
![master](https://cloud.githubusercontent.com/assets/11291849/24301244/64bbd188-10af-11e7-8243-4b1108218ccc.png)
  - `Current PR`
![bad mnt](https://cloud.githubusercontent.com/assets/11291849/24301253/6f45ab60-10af-11e7-8d1a-09d902a54dce.png)

In technical verification, I added a history of texture assignments on each tile.
And there is a difference between this PR and the master.
In this PR, the number of textures assigned by tiles is higher in this PR
Therefore some assignments are unnecessary.
It's a new regression and harshly the origin of the bugs already mentioned.

The input parameters of the texture management mechanism have changed and have not been taken into account to the end, causing new bugs.

> a lot of other work depend on it, it is currently a big blocker for others

To continue working, I advise you to quickly create or rebase yours branches  on `refacto_layer`, if you aren't disturbed by these bugs.

> it does not introduce new bug

Unfortunately, it introduce

> PR atomicity is important

Actually as we can see in  [the presentation of this PR](https://github.com/iTowns/itowns2/pull/239#issue-203158762). It doesn't respect the atomicity rule.

> Can we unblock the situation ?

As seen with @peppsac , the PR remove ancestor #244 should help debug this PR.
Because the remove ancestor PR does not add a bug and simplifies the code.
I recently took into account the @peppsac's reviewing (see commit de209d06d6c8a44f138002ea46bbabaa1ecb7d3b)

There's not much left to do to merge this PR

peppsac commented:
200th comment :) : I think we can merge this one.
It's working ok, the API changes are not all done yet, but I'd prefer to iterate rather than blocking this PR for too long.

gchoqueux commented on src/Process/TiledNodeProcessing.js:
```diff
@@ -0,0 +1,172 @@
+                child.updateMatrixWorld(true);
+                child.OBB().update();
+            }
+            // TODO
```
add comment

gchoqueux commented on src/Core/Commander/Providers/TileProvider.js:
```diff
@@ -43,28 +37,26 @@ TileProvider.prototype.executeCommand = function executeCommand(command) {
-    tile.OBB().update();
 
     // update bbox if node herits texture elevation from parent
     if (tile.material.getElevationLayerLevel() > EMPTY_TEXTURE_ZOOM) {
```
to patch this :
https://github.com/iTowns/itowns2/pull/239/files#diff-e729b37b590009b20e3b637dcc65321fR68

add this 

```js
if (tile.material.getElevationLayerLevel() > EMPTY_TEXTURE_ZOOM) {
    const textureElevation = tile.material.getLayerTextures(l_ELEVATION)[0];
    // need to maximize elevation when delta zoom is important
    const isBigDelta = tile.level - tile.material.getElevationLayerLevel() > 6;

    const { min, max } = !isBigDelta ? OGCWebServiceHelper.ioDXBIL.computeMinMaxElevation(
            textureElevation.image.data,
            SIZE_TEXTURE_TILE, SIZE_TEXTURE_TILE,
            tile.material.offsetScale[0][0]) : {
                min: parent.bbox.minCoordinate.altitude(),
                max: parent.bbox.maxCoordinate.altitude(),
            };

    if (min && max) {
        tile.setBBoxZ(min, max);
    }
}
```

peppsac replied:
Done

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -0,0 +1,134 @@
+        if (node.level < minLevel) {
+            return true;
+        }
+        if (maxLevel <= node.level) {
```
NOTE: minLevel and maxLevel must be modifiable

gchoqueux commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -0,0 +1,248 @@
+}
+
+
+export function initNewNode(geometryLayer, node) {
```
strange function, 
Why there's no lighting property initialisation?

peppsac replied:
I've removed it. And lighting init is in ApiGlobe.

gchoqueux commented:
Otherwise this PR looks very good now

gchoqueux commented:
OK, it's good
Thanks

vpicavet commented:
Thanks for the hard work on this one, it should allow to focus on new progresses now. 
Hope to see next changes merged quicker.
